server, storage: add an in memory-lease history

BramGruneir opened the pull request:
This history is available via the status server and the debug range page.

<img width="742" alt="screen shot 2017-04-16 at 00 43 13" src="https://cloud.githubusercontent.com/assets/1614265/25068940/4ce91df2-2240-11e7-8847-7873e019df1e.png">
<img width="937" alt="screen shot 2017-04-16 at 00 43 05" src="https://cloud.githubusercontent.com/assets/1614265/25068939/4ce81b78-2240-11e7-93f2-824d68076074.png">

BramGruneir committed:
server, storage: add an in memory-lease history

This history is available via the status server and the debug range page.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/14966)
<!-- Reviewable:end -->

petermattis commented:
The lease history table is a bit of an eye chart. I don't have a concrete suggestion at this moment, just a "whoa" reaction.

---

Review status: 0 of 11 files reviewed at latest revision, 5 unresolved discussions.

---

*[pkg/server/debug_range.go, line 230 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14966#-KhtYsZd9k2M94AnQWNT:-KhtYsZd9k2M94AnQWNU:b-ecxrus) ([raw file](https://github.com/cockroachdb/cockroach/blob/1c3b32133949f713d622faabbff57b63307a3648/pkg/server/debug_range.go#L230)):*
> ```Go
> 	}
> 	return l[i].ProposedTS.Less(*l[j].ProposedTS)
> }
> ```

For non-performance critical sorting, I'd recommend `sort.Slice` now. I think this usage definitely falls in that category.

---

*[pkg/storage/lease_history.go, line 24 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14966#-KhtZRjzUadlLW_Ramym:-KhtZRjzUadlLW_Ramyn:b-6pem08) ([raw file](https://github.com/cockroachdb/cockroach/blob/1c3b32133949f713d622faabbff57b63307a3648/pkg/storage/lease_history.go#L24)):*
> ```Go
> 
> // leaseHistoryMaxEntries is the maximum number of lease history entries to
> // store per replica.
> ```

`s/store/maintain/g` to avoid confusion with the normal usage of `store` in this package. 

---

*[pkg/storage/lease_history.go, line 25 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14966#-Kht_Wr3CnRFoH460SQ_:-Kht_Wr3CnRFoH460SQa:bs7b72n) ([raw file](https://github.com/cockroachdb/cockroach/blob/1c3b32133949f713d622faabbff57b63307a3648/pkg/storage/lease_history.go#L25)):*
> ```Go
> // leaseHistoryMaxEntries is the maximum number of lease history entries to
> // store per replica.
> const leaseHistoryMaxEntries = 100
> ```

`sizeof(Lease) == 80` making this not insignificant in terms of memory at large replica counts. I wonder if the lease history should be persisted to some local part of the key space. 

---

*[pkg/storage/lease_history.go, line 32 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14966#-KhtZfejptg8i-eAt6Pv:-KhtZfejptg8i-eAt6Pw:b-mrjnst) ([raw file](https://github.com/cockroachdb/cockroach/blob/1c3b32133949f713d622faabbff57b63307a3648/pkg/storage/lease_history.go#L32)):*
> ```Go
> }
> 
> func newLeaseHistory(clock *hlc.Clock) *leaseHistory {
> ```

The `clock` argument is unused.

---

*[pkg/storage/lease_history.go, line 51 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14966#-KhtZtdaEqed6dpnx6Fx:-KhtZtdaEqed6dpnx6Fy:b-i36pk3) ([raw file](https://github.com/cockroachdb/cockroach/blob/1c3b32133949f713d622faabbff57b63307a3648/pkg/storage/lease_history.go#L51)):*
> ```Go
> 		start = 1
> 	}
> 	lh.history[rangeID] = append(rangeHistory[start:], lease)
> ```

This will allocate a new buffer for `rangeHistory` that is likely larger than `len(rangeHistory)`. I don't think that is what you want. Probably better to just shift the entries around using `copy` as adding a lease entry is rare.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14966#-:-KhtY5vF1KuxXQ_7IqDP:b-d1fa1q)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Showing the second and third timestamps per line as deltas from the start time would help with legibility, I think. 

---

Reviewed 11 of 11 files at r1.
Review status: all files reviewed at latest revision, 11 unresolved discussions.

---

*[pkg/server/debug_range.go, line 230 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14966#-KhtYsZd9k2M94AnQWNT:-KhwNp7nNf5JHqe7Ychf:b-82cz3u) ([raw file](https://github.com/cockroachdb/cockroach/blob/1c3b32133949f713d622faabbff57b63307a3648/pkg/server/debug_range.go#L230)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

For non-performance critical sorting, I'd recommend `sort.Slice` now. I think this usage definitely falls in that category.
</blockquote></details>

This sorting seems to be either redundant or lossy - the history contains leases in the order in which they were applied. I think that always corresponds to `ProposedTS` order, but if it doesn't, we'd want to see that reflected in the output.

---

*[pkg/server/debug_range.go, line 232 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14966#-KhwNfRS_ZttyjMHoZzG:-KhwNfRS_ZttyjMHoZzH:b-4vwn07) ([raw file](https://github.com/cockroachdb/cockroach/blob/1c3b32133949f713d622faabbff57b63307a3648/pkg/server/debug_range.go#L232)):*
> ```Go
> }
> 
> func removeDuplicateLeases(entires leaseSlice) leaseSlice {
> ```

Are duplicate leases a common occurrence? I would think they would be rare enough that we'd want to see them when they happen. 

---

*[pkg/server/debug_range.go, line 234 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14966#-KhwNa9kV29U0aK9yFZW:-KhwNa9kV29U0aK9yFZX:b-xskf83) ([raw file](https://github.com/cockroachdb/cockroach/blob/1c3b32133949f713d622faabbff57b63307a3648/pkg/server/debug_range.go#L234)):*
> ```Go
> func removeDuplicateLeases(entires leaseSlice) leaseSlice {
> 	var result leaseSlice
> 	for _, entry := range entires {
> ```

s/entires/entries/

---

*[pkg/server/debug_range.go, line 631 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14966#-KhwP3azzqEdSb4yTL20:-KhwP3azzqEdSb4yTL21:b-8li2kl) ([raw file](https://github.com/cockroachdb/cockroach/blob/1c3b32133949f713d622faabbff57b63307a3648/pkg/server/debug_range.go#L631)):*
> ```Go
> 	for _, lease := range leaseHistory {
> 		var detail debugLeaseDetails
> 		detail.Replica.Value = fmt.Sprintf("n%d s%d r%d", lease.Replica.NodeID, lease.Replica.StoreID, lease.Replica.ReplicaID)
> ```

I think `r%d` always means "range ID", never "replica ID". When we show the replica ID in abbreviated format, we do it as `"r%d/%d", rangeID, replicaID`.

---

*[pkg/server/status.go, line 627 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14966#-KhwNDlChPdNsLTVzjmq:-KhwNDlChPdNsLTVzjmr:b-tabcdb) ([raw file](https://github.com/cockroachdb/cockroach/blob/1c3b32133949f713d622faabbff57b63307a3648/pkg/server/status.go#L627)):*
> ```Go
> 	}
> 
> 	// TODO(bram): Add in lease info for a range if it is no longer on a store
> ```

Is this really going to be useful? The leases are part of the replicated raft log and should be identical everywhere. We just need the history from one node, the most up-to-date one.

---

*[pkg/storage/lease_history.go, line 25 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14966#-Kht_Wr3CnRFoH460SQ_:-KhwKcKHQ8tfEMSBn8O0:b-k2eov6) ([raw file](https://github.com/cockroachdb/cockroach/blob/1c3b32133949f713d622faabbff57b63307a3648/pkg/storage/lease_history.go#L25)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

`sizeof(Lease) == 80` making this not insignificant in terms of memory at large replica counts. I wonder if the lease history should be persisted to some local part of the key space. 
</blockquote></details>

And we never remove entries from the map, so we'll maintain the history for any range we've ever had a replica of. After going to such great lengths to avoid running a goroutine (4kb) per replica, keeping this much data in memory seems like a bad idea. Could we do something like the new timestamp cache to set a global limit instead of a per-replica one?

---

*[pkg/storage/lease_history.go, line 45 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14966#-KhwLc8zxXcZBin1y6E5:-KhwLc8zxXcZBin1y6E6:b4aax8o) ([raw file](https://github.com/cockroachdb/cockroach/blob/1c3b32133949f713d622faabbff57b63307a3648/pkg/storage/lease_history.go#L45)):*
> ```Go
> 	rangeHistory, ok := lh.history[rangeID]
> 	if !ok {
> 		rangeHistory = make([]roachpb.Lease, 0, leaseHistoryMaxEntries)
> ```

We shouldn't eagerly allocate the buffer here. With epoch-based leases, changes are rare and many ranges won't need the whole buffer.

---

*[pkg/storage/lease_history.go, line 51 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14966#-KhtZtdaEqed6dpnx6Fx:-KhwMFlHYmBRSdkhdMpL:b-qwec8) ([raw file](https://github.com/cockroachdb/cockroach/blob/1c3b32133949f713d622faabbff57b63307a3648/pkg/storage/lease_history.go#L51)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

This will allocate a new buffer for `rangeHistory` that is likely larger than `len(rangeHistory)`. I don't think that is what you want. Probably better to just shift the entries around using `copy` as adding a lease entry is rare.
</blockquote></details>

But mutating the array with `copy()` will be thread-unsafe with this implementation of `Get`. 

---

*[pkg/storage/replica_command.go, line 2037 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14966#-KhwKNV7vsRlnXd77aL0:-KhwKNV7vsRlnXd77aL1:beyn64e) ([raw file](https://github.com/cockroachdb/cockroach/blob/1c3b32133949f713d622faabbff57b63307a3648/pkg/storage/replica_command.go#L2037)):*
> ```Go
> 	}
> 	// Mark the new lease in the store's lease history.
> 	rec.repl.store.leaseHistory.Add(desc.RangeID, lease)
> ```

No side effects are allowed in replica_command.go. This needs to go in replica_proposal.go. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14966#-:-KhwOcDgjFbGK1dtYKxk:bxtgsv7)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
I agree, it's quite ugly.  This was my 3rd revision of it.  You don't want to see the earlier drafts.  As always, I'd be more than happy to make it easier on the eyes.

I've replace the 2nd and 3rd timestamps with deltas.  Much nicer to read.

![Screen Shot 2017-04-17 at 14.41.54.png](https://files.reviewableusercontent.io/images/reviewable/1614265/isKRflo5IGf2zWmooY1NhdWF/Screen%20Shot%202017-04-17%20at%2014.41.54.png)
![Screen Shot 2017-04-17 at 14.41.48.png](https://files.reviewableusercontent.io/images/reviewable/1614265/k_JXXyOtl1XjtEUcHP9Yn9Mw/Screen%20Shot%202017-04-17%20at%2014.41.48.png)

---

Review status: 6 of 11 files reviewed at latest revision, 11 unresolved discussions, some commit checks pending.

---

*[pkg/server/debug_range.go, line 230 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14966#-KhtYsZd9k2M94AnQWNT:-KhwSUdOFP_neWOPRxhg:bnpk1vn) ([raw file](https://github.com/cockroachdb/cockroach/blob/1c3b32133949f713d622faabbff57b63307a3648/pkg/server/debug_range.go#L230)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

This sorting seems to be either redundant or lossy - the history contains leases in the order in which they were applied. I think that always corresponds to `ProposedTS` order, but if it doesn't, we'd want to see that reflected in the output.
</blockquote></details>

It's not redundant since we're pulling leases form all replicas to create the history.  I could do a merge sort instead, but that seemed excessive.  

I'd sort by just ProposedTS, but I wanted to deal with the uber-slim change that the proposedTS is nil.  

The rest of the sorts in this file use the old style, so I figured it was easy enough to follow suit with this new one.  If you feel strongly, I'll change them all, but I don't see a compelling reason to do so.  

---

*[pkg/server/debug_range.go, line 232 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14966#-KhwNfRS_ZttyjMHoZzG:-KhwSWUhZvSnCGzjCtgK:b-ftiosq) ([raw file](https://github.com/cockroachdb/cockroach/blob/1c3b32133949f713d622faabbff57b63307a3648/pkg/server/debug_range.go#L232)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Are duplicate leases a common occurrence? I would think they would be rare enough that we'd want to see them when they happen. 
</blockquote></details>

These are not duplicate leases in the sense that two leases are applied at once, this grabs the leases from all stores with a range's replica.  And each store contains a lease history of all applied leases.  So this is just deduping those duplicates.  

---

*[pkg/server/debug_range.go, line 234 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14966#-KhwNa9kV29U0aK9yFZW:-KhwTtoIqSaEuvI8EK5n:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1c3b32133949f713d622faabbff57b63307a3648/pkg/server/debug_range.go#L234)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

s/entires/entries/
</blockquote></details>

Done.

---

*[pkg/server/debug_range.go, line 631 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14966#-KhwP3azzqEdSb4yTL20:-KhwUDCdGMItIesUO0au:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1c3b32133949f713d622faabbff57b63307a3648/pkg/server/debug_range.go#L631)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

I think `r%d` always means "range ID", never "replica ID". When we show the replica ID in abbreviated format, we do it as `"r%d/%d", rangeID, replicaID`.
</blockquote></details>

Done.

---

*[pkg/server/status.go, line 627 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14966#-KhwNDlChPdNsLTVzjmq:-KhwUGf-bkAvTp86lEPN:b-73iq5q) ([raw file](https://github.com/cockroachdb/cockroach/blob/1c3b32133949f713d622faabbff57b63307a3648/pkg/server/status.go#L627)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Is this really going to be useful? The leases are part of the replicated raft log and should be identical everywhere. We just need the history from one node, the most up-to-date one.
</blockquote></details>

I don't think looking through the raft log is the right way to go here.  I could add a lease history to the range local keyspace, but I don't know if that's really needed.  I don't think that the lease history is important enough for that, but it wouldn't be too hard to do.  

The point of this is that if if we use the current in-memory lease history and all three replicas jump move to new stores, the in-memory lease history would be lost unless you looked in the stores that no longer have a replica.

Mind you, this is a bit of an edge case, but I've tested a good amount of zone config changes where this is exactly the behaviour I was trying to gain insight into.

---

*[pkg/storage/lease_history.go, line 24 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14966#-KhtZRjzUadlLW_Ramym:-KhwVXCFTAwvhme6b13U:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1c3b32133949f713d622faabbff57b63307a3648/pkg/storage/lease_history.go#L24)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

`s/store/maintain/g` to avoid confusion with the normal usage of `store` in this package. 
</blockquote></details>

Done.

---

*[pkg/storage/lease_history.go, line 25 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14966#-Kht_Wr3CnRFoH460SQ_:-KhwVgiNL6jAlQiJIyBe:bfj1ssd) ([raw file](https://github.com/cockroachdb/cockroach/blob/1c3b32133949f713d622faabbff57b63307a3648/pkg/storage/lease_history.go#L25)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

And we never remove entries from the map, so we'll maintain the history for any range we've ever had a replica of. After going to such great lengths to avoid running a goroutine (4kb) per replica, keeping this much data in memory seems like a bad idea. Could we do something like the new timestamp cache to set a global limit instead of a per-replica one?
</blockquote></details>

Sure, I can maintain a global limit instead.  Done.

---

*[pkg/storage/lease_history.go, line 32 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14966#-KhtZfejptg8i-eAt6Pv:-KhwK4cx5WXXheY5jl2Y:b-5d3uqj) ([raw file](https://github.com/cockroachdb/cockroach/blob/1c3b32133949f713d622faabbff57b63307a3648/pkg/storage/lease_history.go#L32)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

The `clock` argument is unused.
</blockquote></details>

Removed, was a vestige of an earlier draft.

---

*[pkg/storage/lease_history.go, line 45 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14966#-KhwLc8zxXcZBin1y6E5:-Khw_IkAwKLwO94xV3dI:b-eztt4g) ([raw file](https://github.com/cockroachdb/cockroach/blob/1c3b32133949f713d622faabbff57b63307a3648/pkg/storage/lease_history.go#L45)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

We shouldn't eagerly allocate the buffer here. With epoch-based leases, changes are rare and many ranges won't need the whole buffer.
</blockquote></details>

I'm going to leave this in since I've made the change to use a circular buffer instead.

---

*[pkg/storage/lease_history.go, line 51 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14966#-KhtZtdaEqed6dpnx6Fx:-Khw__LkBZWkUbAt6FTY:bg1qmon) ([raw file](https://github.com/cockroachdb/cockroach/blob/1c3b32133949f713d622faabbff57b63307a3648/pkg/storage/lease_history.go#L51)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

But mutating the array with `copy()` will be thread-unsafe with this implementation of `Get`. 
</blockquote></details>

I've changed this implementation to a circular buffer instead.  It allows us to completely control the amount of space used.

---

*[pkg/storage/replica_command.go, line 2037 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14966#-KhwKNV7vsRlnXd77aL0:-Khx5E5CQ62YGVkMPgHj:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1c3b32133949f713d622faabbff57b63307a3648/pkg/storage/replica_command.go#L2037)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

No side effects are allowed in replica_command.go. This needs to go in replica_proposal.go. 
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14966#-:-KhwJKR1laywZ8u9kagy:b-2uegw4)*
<!-- Sent from Reviewable.io -->

a-robinson commented:
Reviewed 10 of 11 files at r1, 5 of 6 files at r2.
Review status: 10 of 11 files reviewed at latest revision, 12 unresolved discussions, some commit checks failed.

---

*[pkg/server/debug_range.go, line 230 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14966#-KhtYsZd9k2M94AnQWNT:-Khx1GwnhPQAiw_M2BPM:b650r6v) ([raw file](https://github.com/cockroachdb/cockroach/blob/1c3b32133949f713d622faabbff57b63307a3648/pkg/server/debug_range.go#L230)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

It's not redundant since we're pulling leases form all replicas to create the history.  I could do a merge sort instead, but that seemed excessive.  

I'd sort by just ProposedTS, but I wanted to deal with the uber-slim change that the proposedTS is nil.  

The rest of the sorts in this file use the old style, so I figured it was easy enough to follow suit with this new one.  If you feel strongly, I'll change them all, but I don't see a compelling reason to do so.  
</blockquote></details>

I'm curious about the choice of `ProposedTS` for sorting as well. Mind adding a comment with the reasoning for using that rather than just `Start`, which appears to be guaranteed to be monotonic?

---

*[pkg/server/debug_range.go, line 631 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14966#-KhwP3azzqEdSb4yTL20:-KhxGhv9ugrlYe4tF9di:b-m6e74x) ([raw file](https://github.com/cockroachdb/cockroach/blob/1c3b32133949f713d622faabbff57b63307a3648/pkg/server/debug_range.go#L631)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

nit, but I might switch the spaces to commas to more closely match our log format

---

*[pkg/storage/lease_history.go, line 25 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14966#-Kht_Wr3CnRFoH460SQ_:-KhxJOwLUrfahPWhNbEK:bjmxwo7) ([raw file](https://github.com/cockroachdb/cockroach/blob/1c3b32133949f713d622faabbff57b63307a3648/pkg/storage/lease_history.go#L25)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

Sure, I can maintain a global limit instead.  Done.
</blockquote></details>

100 isn't going to be particularly useful as a global limit. Also, no matter what limit you choose it'll get pretty quickly overwhelmed by the expiration-based leases if we're creating a new entry for those every few seconds.

---

*[pkg/storage/replica_proposal.go, line 467 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14966#-KhxI90IxjAwI7GVw1me:-KhxI90IxjAwI7GVw1mf:btnj0ax) ([raw file](https://github.com/cockroachdb/cockroach/blob/c8f2c18f221670a429b86753af188654fba89e78/pkg/storage/replica_proposal.go#L467)):*
> ```Go
> 
> 	// Mark the new lease in the store's lease history.
> 	if newLease != nil {
> ```

I don't think `newLease` is allowed to be nil here.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14966)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
Review status: 9 of 11 files reviewed at latest revision, 12 unresolved discussions.

---

*[pkg/server/debug_range.go, line 230 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14966#-KhtYsZd9k2M94AnQWNT:-KhxSk1JTh4KaIAIhGRD:bwhbbr6) ([raw file](https://github.com/cockroachdb/cockroach/blob/1c3b32133949f713d622faabbff57b63307a3648/pkg/server/debug_range.go#L230)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

I'm curious about the choice of `ProposedTS` for sorting as well. Mind adding a comment with the reasoning for using that rather than just `Start`, which appears to be guaranteed to be monotonic?
</blockquote></details>

It seemed like the most appropriate, since it gets set when the lease is actually being considered.  But sorting by start is cool too.

Is start guaranteed to be monotonic?  I'm pretty sure proposed is.

---

*[pkg/server/debug_range.go, line 631 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14966#-KhwP3azzqEdSb4yTL20:-KhxRvY6Xpq1SAierHJy:b-19kw1d) ([raw file](https://github.com/cockroachdb/cockroach/blob/1c3b32133949f713d622faabbff57b63307a3648/pkg/server/debug_range.go#L631)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

nit, but I might switch the spaces to commas to more closely match our log format
</blockquote></details>

I'm going to leave it as is.  In this case, I think it's more readable.  Unless it's a use case to copy and then grep the logs from this page. 

---

*[pkg/storage/lease_history.go, line 25 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14966#-Kht_Wr3CnRFoH460SQ_:-KhxMBdn7QNNzo1toQyw:b-ykjads) ([raw file](https://github.com/cockroachdb/cockroach/blob/1c3b32133949f713d622faabbff57b63307a3648/pkg/storage/lease_history.go#L25)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

100 isn't going to be particularly useful as a global limit. Also, no matter what limit you choose it'll get pretty quickly overwhelmed by the expiration-based leases if we're creating a new entry for those every few seconds.
</blockquote></details>

I just picked 100 of out nowhere.  But you're right, for epoch based leases, if the lease has been stable for a long time, it will get pushed out.  I could a second buffer, so we would have one for epoch and one for expiration.  That would solve that problem easily enough.

Thoughts?

---

*[pkg/storage/replica_proposal.go, line 467 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14966#-KhxI90IxjAwI7GVw1me:-KhxMQ0oulrTuZ41rmKy:b-ruuro1) ([raw file](https://github.com/cockroachdb/cockroach/blob/c8f2c18f221670a429b86753af188654fba89e78/pkg/storage/replica_proposal.go#L467)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

I don't think `newLease` is allowed to be nil here.
</blockquote></details>

It's not, this check is unneeded.  Removed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14966)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Hmm, for the expiration-based ranges it's weird that the leases are proposed after their start time when they refresh (and shouldn't there be at least some cases where the delta is negative?). Maybe Proposed is the timestamp that deserves to be shown in full, since it's the sort criterion. 

And how did the epoch-based lease get proposed 4 hours after it started? 

---

Reviewed 4 of 6 files at r2, 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 11 unresolved discussions, some commit checks failed.

---

*[pkg/server/debug_range.go, line 230 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14966#-KhtYsZd9k2M94AnQWNT:-KhxnhixfrYZn_uXyuBE:bd21gz4) ([raw file](https://github.com/cockroachdb/cockroach/blob/1c3b32133949f713d622faabbff57b63307a3648/pkg/server/debug_range.go#L230)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

It seemed like the most appropriate, since it gets set when the lease is actually being considered.  But sorting by start is cool too.

Is start guaranteed to be monotonic?  I'm pretty sure proposed is.
</blockquote></details>

Start never goes backwards, but it may stay the same for consecutive leases (as can be seen in the screenshot above) 

---

*[pkg/server/debug_range.go, line 232 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14966#-KhwNfRS_ZttyjMHoZzG:-KhxmLba7OeLg1wWhm0n:b-qf6169) ([raw file](https://github.com/cockroachdb/cockroach/blob/1c3b32133949f713d622faabbff57b63307a3648/pkg/server/debug_range.go#L232)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

These are not duplicate leases in the sense that two leases are applied at once, this grabs the leases from all stores with a range's replica.  And each store contains a lease history of all applied leases.  So this is just deduping those duplicates.  
</blockquote></details>

OK. I think instead of sorting and merging the lists from different nodes, we should just discard all but the latest, since everything from the others will be redundant. 

---

*[pkg/server/status.go, line 627 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14966#-KhwNDlChPdNsLTVzjmq:-Khxme8ncb8P9RdsMUTK:bis69zo) ([raw file](https://github.com/cockroachdb/cockroach/blob/1c3b32133949f713d622faabbff57b63307a3648/pkg/server/status.go#L627)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

I don't think looking through the raft log is the right way to go here.  I could add a lease history to the range local keyspace, but I don't know if that's really needed.  I don't think that the lease history is important enough for that, but it wouldn't be too hard to do.  

The point of this is that if if we use the current in-memory lease history and all three replicas jump move to new stores, the in-memory lease history would be lost unless you looked in the stores that no longer have a replica.

Mind you, this is a bit of an edge case, but I've tested a good amount of zone config changes where this is exactly the behaviour I was trying to gain insight into.
</blockquote></details>

If that's a goal then maybe we should figure out how to store this in a global/persistent way instead of piecing together incomplete in-memory records from different nodes. 

---

*[pkg/storage/lease_history.go, line 25 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14966#-Kht_Wr3CnRFoH460SQ_:-Khxm4wTXHKfS63YMcn1:b-u6ql2w) ([raw file](https://github.com/cockroachdb/cockroach/blob/1c3b32133949f713d622faabbff57b63307a3648/pkg/storage/lease_history.go#L25)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

I just picked 100 of out nowhere.  But you're right, for epoch based leases, if the lease has been stable for a long time, it will get pushed out.  I could a second buffer, so we would have one for epoch and one for expiration.  That would solve that problem easily enough.

Thoughts?
</blockquote></details>

I think one buffer for epoch and one for expiration ranges makes sense. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14966#-:-KhxnISsZXrkpyIOmXCc:b-fqdcn8)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: all files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[pkg/server/status.go, line 627 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14966#-KhwNDlChPdNsLTVzjmq:-KhxodB2UNRo9eWNBWmL:b-l7vlu3) ([raw file](https://github.com/cockroachdb/cockroach/blob/1c3b32133949f713d622faabbff57b63307a3648/pkg/server/status.go#L627)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

If that's a goal then maybe we should figure out how to store this in a global/persistent way instead of piecing together incomplete in-memory records from different nodes. 
</blockquote></details>

https://github.com/cockroachdb/cockroach/issues/13210

---

*[pkg/storage/lease_history.go, line 25 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14966#-Kht_Wr3CnRFoH460SQ_:-Khxngg6NsVHSqIQMwr0:b9gisor) ([raw file](https://github.com/cockroachdb/cockroach/blob/1c3b32133949f713d622faabbff57b63307a3648/pkg/storage/lease_history.go#L25)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

I just picked 100 of out nowhere.  But you're right, for epoch based leases, if the lease has been stable for a long time, it will get pushed out.  I could a second buffer, so we would have one for epoch and one for expiration.  That would solve that problem easily enough.

Thoughts?
</blockquote></details>

Perhaps have separate `history` storage for epoch-based and expiration-based leases. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14966)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
No more sorting anymore, one source of truth for the leases means that they are pre-sorted based on when they were added.  This is still showing relative values compared to start.  Also, I added deltas from the previous lease, so that should make this even easier on the eyes.  

Not shown in the screenshots, but if you hover the mouse over any of the times, you'll see the raw timestamp and the raw converted date.  This includes start, proposed and expiration.

As far as why these number are strange, we'll have to see what a real cluster looks like, since I've been starting/stopping roachdemo over and over.

![Screen Shot 2017-04-18 at 15.04.58.png](https://files.reviewableusercontent.io/images/reviewable/1614265/6LslVMxzZun96LhhAewH3NzX/Screen%20Shot%202017-04-18%20at%2015.04.58.png)
![Screen Shot 2017-04-18 at 15.04.50.png](https://files.reviewableusercontent.io/images/reviewable/1614265/cec6GIG2nOkSJ0BZsIJV4Dw-/Screen%20Shot%202017-04-18%20at%2015.04.50.png)

---

Review status: 8 of 11 files reviewed at latest revision, 11 unresolved discussions.

---

*[pkg/server/debug_range.go, line 230 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14966#-KhtYsZd9k2M94AnQWNT:-Ki17c8BznlOleN8ScOE:bptgx5e) ([raw file](https://github.com/cockroachdb/cockroach/blob/1c3b32133949f713d622faabbff57b63307a3648/pkg/server/debug_range.go#L230)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Start never goes backwards, but it may stay the same for consecutive leases (as can be seen in the screenshot above) 
</blockquote></details>

I've removed the sorting since we're only pulling from a single source now, they're always already sorted.

---

*[pkg/server/debug_range.go, line 232 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14966#-KhwNfRS_ZttyjMHoZzG:-Ki1arkC1SQrf_NSPGIR:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1c3b32133949f713d622faabbff57b63307a3648/pkg/server/debug_range.go#L232)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

OK. I think instead of sorting and merging the lists from different nodes, we should just discard all but the latest, since everything from the others will be redundant. 
</blockquote></details>

Done.

---

*[pkg/server/status.go, line 627 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14966#-KhwNDlChPdNsLTVzjmq:-Ki14lJbDTeC3NGbc3jx:b9a5uzu) ([raw file](https://github.com/cockroachdb/cockroach/blob/1c3b32133949f713d622faabbff57b63307a3648/pkg/server/status.go#L627)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

https://github.com/cockroachdb/cockroach/issues/13210
</blockquote></details>

I don't think now is the time to start adding in new writes as part of the lease process.  1.1?

---

*[pkg/storage/lease_history.go, line 25 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14966#-Kht_Wr3CnRFoH460SQ_:-Ki14uPWhCDnA6ecPwof:b-g01e9e) ([raw file](https://github.com/cockroachdb/cockroach/blob/1c3b32133949f713d622faabbff57b63307a3648/pkg/storage/lease_history.go#L25)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Perhaps have separate `history` storage for epoch-based and expiration-based leases. 
</blockquote></details>

Done, and I made the max value an env variable so we can request longer histories if we see a need to. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14966#-:-Ki1b4w3V3cdEYwXGkye:b8orzv)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
One more small change, I added the node id to the epoch so that it's clear the epoch is not moving backwards.

![Screen Shot 2017-04-18 at 16.10.42.png](https://files.reviewableusercontent.io/images/reviewable/1614265/esS1nBvism42j4OgNxOfVWcz/Screen%20Shot%202017-04-18%20at%2016.10.42.png)

---

Review status: 6 of 12 files reviewed at latest revision, 10 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14966#-:-Ki1gGuo5JhC-C01f1VK:b-3zbqsf)*
<!-- Sent from Reviewable.io -->

a-robinson commented:
The Proposed column for epoch-based leases in your latest screenshot confuses me, and I'm not sure what I'm supposed to be getting out of it. Can you explain? Is it adding any value?

---

Reviewed 1 of 6 files at r2, 6 of 6 files at r4.
Review status: all files reviewed at latest revision, 12 unresolved discussions, some commit checks failed.

---

*[pkg/server/debug_range.go, line 236 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/14966#-Ki6ASuI4OMuKjlWZJpU:-Ki6ASuJ1t8wSvmsnViF:b-sausbw) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d572f40c77bacd9910f8b5ae062720a2463c9ec/pkg/server/debug_range.go#L236)):*
> ```Go
> 	HeaderFakeStoreID roachpb.StoreID
> 	LeaseHistory      []debugLeaseDetails
> 	LeaseEpoch        bool // true of epoch based, false of expiration based
> ```

s/of/if

---

*[pkg/server/debug_range.go, line 316 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/14966#-Ki6B1jm4mROyIbTuplW:-Ki6B1jm4mROyIbTuplX:b-7dyyl5) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d572f40c77bacd9910f8b5ae062720a2463c9ec/pkg/server/debug_range.go#L316)):*
> ```Go
> 	// leaderlessLatestTermInfo is used when there is no leader but we want to
> 	// display some details about one of the replicas.
> 	var leaderlessLatestTermInfo serverpb.RangeInfo
> ```

Just calling it `latestTermInfo` would make things clearer, to me at least, since it ends up getting used as the source of truth later

---

*[pkg/storage/lease_history.go, line 25 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14966#-Kht_Wr3CnRFoH460SQ_:-Ki6BslL5vn3TfhWp0GT:bt4dm8x) ([raw file](https://github.com/cockroachdb/cockroach/blob/1c3b32133949f713d622faabbff57b63307a3648/pkg/storage/lease_history.go#L25)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

Done, and I made the max value an env variable so we can request longer histories if we see a need to. 
</blockquote></details>

100 won't even be enough to have the most recent epoch-based lease from each range on a reasonably sized node. Given that each lease is 80 bytes, we could reasonably keep tens of thousands of these around. 50k epoch-based leases only would take 4MB of memory.

I don't know that we'd want to pay that much memory to keep so many expiration-based leases around, though.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14966#-:-Ki6D26R4IjLJybjtYPK:b-q4rd9l)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
"Proposed" and "Proposed Delta" are both deltas; we need to be clear about what these are relative to. 

I think that since start times are reused but proposed times (mostly) aren't, maybe the proposed time should be first (the one reused propose time in the last screenshot is strange, and might be a bug somewhere). In fact, since this list is filtered down to leases that were actually applied, I'm not sure that start and expiration times are useful at all. What if the table was "Replica", "Proposed", "Delta" (proposed - previous proposed), "Epoch" (if applicable), with "Start" and "Expiration" only available on hover?

---

Reviewed 6 of 6 files at r4.
Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[pkg/storage/lease_history.go, line 25 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14966#-Kht_Wr3CnRFoH460SQ_:-Ki6_4uu5cN7dQV9nU58:bnaxv14) ([raw file](https://github.com/cockroachdb/cockroach/blob/1c3b32133949f713d622faabbff57b63307a3648/pkg/storage/lease_history.go#L25)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

100 won't even be enough to have the most recent epoch-based lease from each range on a reasonably sized node. Given that each lease is 80 bytes, we could reasonably keep tens of thousands of these around. 50k epoch-based leases only would take 4MB of memory.

I don't know that we'd want to pay that much memory to keep so many expiration-based leases around, though.
</blockquote></details>

Yeah, 100 seems too small to be useful. It should be at least 10k. 

---

*[pkg/storage/lease_history.go, line 92 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/14966#-Ki6Zi3sl_MT2qWVUBo5:-Ki6Zi3sl_MT2qWVUBo6:bso9lpm) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d572f40c77bacd9910f8b5ae062720a2463c9ec/pkg/storage/lease_history.go#L92)):*
> ```Go
> 		return results
> 	}
> 	for i := lh.expirationIndex; i < lh.expirationIndex+leaseHistoryMaxEntries; i++ {
> ```

This will get slow if the cache size is increased, but I guess since that's an opt-in thing we don't need to worry about it for now.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14966#-:-Ki6TUYQmlM1FCqh0NPQ:b-wrvmpv)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
I'm cool with changing this up, but hiding start and epoch until hover is a bit of fancy js that I'd rather not add to a static page right now.  For hovers, I'm just using the title attribute for the cells.

How does this sound:
For expiration:
replica - proposed - proposed delta (from the last proposed) - start - start delta (to the last start) - expiration (delta from the start)

For Epoch:
replica - proposed - proposed delta (from the last proposed) - start - start delta (to the last start) - epoch

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14966#-:-Ki775USpufs6baApiQ9:b99i1up)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
Ok, I've done a good amount of refactoring to this.  

- Lease Histories can now be enabled via the env flag "COCKROACH_LEASE_HISTORY" and they are defaulted to off.
- Histories are now stored on the replica instead of the store.
- The output has been updated again for readability.

The original goal of using the store was there for two reasons:
1) to keep the histories during heavy rebalancing, (all 3 replicas have moved to new locations)
2) tightly control memory usage

Since this is disabled by default, 2 is no longer an issue.
Keeping the history after a replica has left a store isn't really a needed use case right now. If it is needed, we should instead move the history into physical range local storage instead of keeping them only in memory.

---

Review status: 4 of 13 files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[pkg/server/debug_range.go, line 236 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/14966#-Ki6ASuI4OMuKjlWZJpU:-Ki6UV0JDdGzWYJz52UW:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d572f40c77bacd9910f8b5ae062720a2463c9ec/pkg/server/debug_range.go#L236)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

s/of/if
</blockquote></details>

Done.

---

*[pkg/server/debug_range.go, line 316 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/14966#-Ki6B1jm4mROyIbTuplW:-Ki6T9hsyqizOtXgCAvv:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d572f40c77bacd9910f8b5ae062720a2463c9ec/pkg/server/debug_range.go#L316)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

Just calling it `latestTermInfo` would make things clearer, to me at least, since it ends up getting used as the source of truth later
</blockquote></details>

Done.

---

*[pkg/storage/lease_history.go, line 25 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14966#-Kht_Wr3CnRFoH460SQ_:-Ki6TrGrOmNcfVz0YWbT:b-h92pkt) ([raw file](https://github.com/cockroachdb/cockroach/blob/1c3b32133949f713d622faabbff57b63307a3648/pkg/storage/lease_history.go#L25)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Yeah, 100 seems too small to be useful. It should be at least 10k. 
</blockquote></details>

So I've set the default to 100 per replica since this is now opt in to enable it.  If we're seeing a significant impact on memory, it can be reduce easily enough. 

---

*[pkg/storage/lease_history.go, line 92 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/14966#-Ki6Zi3sl_MT2qWVUBo5:-Ki6aJSiF50zvJRWbHHj:b-suk3k1) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d572f40c77bacd9910f8b5ae062720a2463c9ec/pkg/storage/lease_history.go#L92)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

This will get slow if the cache size is increased, but I guess since that's an opt-in thing we don't need to worry about it for now.
</blockquote></details>

Changed this to store the history within the replica so this shouldn't be an issue.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14966#-:-Ki8gD3EaMR-DWHanY3V:buu0wnb)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
Here are the latest screenshots:

![Screen Shot 2017-04-20 at 00.37.44.png](https://files.reviewableusercontent.io/images/reviewable/1614265/-Rnoza_cU6ym0oEi2asIvOkP/Screen%20Shot%202017-04-20%20at%2000.37.44.png)
![Screen Shot 2017-04-20 at 00.37.38.png](https://files.reviewableusercontent.io/images/reviewable/1614265/FWZxJRCCjB1zCyy0x4UwjPGH/Screen%20Shot%202017-04-20%20at%2000.37.38.png)

---

Review status: 4 of 13 files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14966#-:-Ki8jfSQVYiHOJj6f7cf:btatudw)*
<!-- Sent from Reviewable.io -->

petermattis commented:
I thought expiration-based leases were renewed every 9 seconds or so. Why does the lease history for `r1` only show a new entry once per minute?

---

Review status: 4 of 13 files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14966#-:-KiAZ6qCJHnCoxkwOLfi:bm2kkav)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
In this case, it's due to the range being inactive and there's a 1 min check on it to renew leases.

Per in person comments, I've changed the expiration to be time since proposed instead of time since start to and I've moved it between proposed delta and start.

![Screen Shot 2017-04-20 at 11.08.55.png](https://files.reviewableusercontent.io/images/reviewable/1614265/ymSh9wFDkgt3m2sMVqfUNZeV/Screen%20Shot%202017-04-20%20at%2011.08.55.png)

---

Review status: 4 of 13 files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14966#-:-KiAsEAC6n-iu1zOSOwb:b-9lmxwv)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 5 of 10 files at r5, 5 of 6 files at r6.
Review status: all files reviewed at latest revision, 11 unresolved discussions, all commit checks successful.

---

*[pkg/storage/lease_history.go, line 61 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/14966#-KiBEfByMjRnXLd4IDf2:-KiBEfByMjRnXLd4IDf3:b-a6cp69) ([raw file](https://github.com/cockroachdb/cockroach/blob/39ca3da1d98208f7e034fbc2226903d09ee25c57/pkg/storage/lease_history.go#L61)):*
> ```Go
> 	}
> 	if len(lh.history) < leaseHistoryMaxEntries {
> 		return lh.history
> ```

This is thread-unsafe; we need to always copy the list.

---

*[pkg/storage/replica.go, line 101 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/14966#-KiBErCFFniIoUja6HQx:-KiBErCFFniIoUja6HQy:b-xpwubr) ([raw file](https://github.com/cockroachdb/cockroach/blob/39ca3da1d98208f7e034fbc2226903d09ee25c57/pkg/storage/replica.go#L101)):*
> ```Go
> // leaseHistoryEnabled controls if replica lease histories are enabled. Turning
> // them on will encour some memory usage.
> var leaseHistoryEnabled = envutil.EnvOrDefaultBool("COCKROACH_LEASE_HISTORY", false)
> ```

It seems better to make this a number defaulting to zero so you can control the memory budget too. 

---

*[pkg/storage/replica.go, line 4869 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/14966#-KiBEyyZwLBSR7aC2IPi:-KiBEyyZwLBSR7aC2IPj:b-ddy7pp) ([raw file](https://github.com/cockroachdb/cockroach/blob/39ca3da1d98208f7e034fbc2226903d09ee25c57/pkg/storage/replica.go#L4869)):*
> ```Go
> func (r *Replica) GetLeaseHistory() []roachpb.Lease {
> 	if r.leaseHistory == nil {
> 		return []roachpb.Lease{}
> ```

This can be nil. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14966#-:-KiBFTtrUxrRGKWqzqcj:bnfp4nl)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
Review status: 10 of 14 files reviewed at latest revision, 11 unresolved discussions, some commit checks pending.

---

*[pkg/storage/lease_history.go, line 61 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/14966#-KiBEfByMjRnXLd4IDf2:-KiBlRGhuVauLoOoNAxQ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/39ca3da1d98208f7e034fbc2226903d09ee25c57/pkg/storage/lease_history.go#L61)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

This is thread-unsafe; we need to always copy the list.
</blockquote></details>

Done.

---

*[pkg/storage/replica.go, line 101 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/14966#-KiBErCFFniIoUja6HQx:-KiBk6Y92lqLbVVEwQKq:b-1eh3wi) ([raw file](https://github.com/cockroachdb/cockroach/blob/39ca3da1d98208f7e034fbc2226903d09ee25c57/pkg/storage/replica.go#L101)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

It seems better to make this a number defaulting to zero so you can control the memory budget too. 
</blockquote></details>

Sure, done.

---

*[pkg/storage/replica.go, line 4869 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/14966#-KiBEyyZwLBSR7aC2IPi:-KiBlWZ5GLy7n3EoYm6Y:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/39ca3da1d98208f7e034fbc2226903d09ee25c57/pkg/storage/replica.go#L4869)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

This can be nil. 
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14966)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
Can someone take one last quick look?  I'd like to get this merged before I have to rebase and updated embedded again.

---

Review status: 6 of 14 files reviewed at latest revision, 11 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14966#-:-KiHFbm1uQsDxVdUvZp-:b-tiz7fq)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Reviewed 9 of 9 files at r7.
Review status: all files reviewed at latest revision, 13 unresolved discussions, all commit checks successful.

---

*[pkg/storage/lease_history.go, line 23 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/14966#-KiMXzamCmZM-MsrU0c5:-KiMXzamCmZM-MsrU0c6:b-vxe1zv) ([raw file](https://github.com/cockroachdb/cockroach/blob/cbc2d6aa2b992a32156df082047dc1c2aa0be17b/pkg/storage/lease_history.go#L23)):*
> ```Go
> )
> 
> const defaultLeaseHistoryMaxEntries = 0
> ```

This constant isn't used anywhere else, so just inline it into the env var definition. 

---

*[pkg/storage/lease_history.go, line 69 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/14966#-KiMYKNZQ4s1lCTMZeIX:-KiMYKNZQ4s1lCTMZeIY:b-ib4aru) ([raw file](https://github.com/cockroachdb/cockroach/blob/cbc2d6aa2b992a32156df082047dc1c2aa0be17b/pkg/storage/lease_history.go#L69)):*
> ```Go
> 		return lh.history
> 	}
> 	return append(lh.history[lh.index:], lh.history[:lh.index]...)
> ```

Doesn't this share the backing array? It doesn't look like you're guaranteed to get a fresh allocation here (especially when `lh.index` is zero).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14966)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
Review status: 13 of 14 files reviewed at latest revision, 13 unresolved discussions.

---

*[pkg/storage/lease_history.go, line 23 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/14966#-KiMXzamCmZM-MsrU0c5:-KiNnltcPk6YZy6_B6oZ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/cbc2d6aa2b992a32156df082047dc1c2aa0be17b/pkg/storage/lease_history.go#L23)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

This constant isn't used anywhere else, so just inline it into the env var definition. 
</blockquote></details>

Done.

---

*[pkg/storage/lease_history.go, line 69 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/14966#-KiMYKNZQ4s1lCTMZeIX:-KiNnmitlk9xkb1bA32h:b-yzjkh1) ([raw file](https://github.com/cockroachdb/cockroach/blob/cbc2d6aa2b992a32156df082047dc1c2aa0be17b/pkg/storage/lease_history.go#L69)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Doesn't this share the backing array? It doesn't look like you're guaranteed to get a fresh allocation here (especially when `lh.index` is zero).
</blockquote></details>

Ah, it does,  I fixed the outright copy case, but not this one.  Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14966)*
<!-- Sent from Reviewable.io -->

a-robinson commented:
Now that we're back to storing the histories per-replica, can you explain somewhere what the intended usage pattern is going to be? Will this just be enabled all the time on our prod clusters? The fact that we can't keep it enabled by default is a bit of a letdown, since it isn't the sort of thing for which you can stop and restart a broken cluster and get useful information from after the restart.

---

Reviewed 1 of 10 files at r5, 1 of 6 files at r6, 2 of 9 files at r7.
Review status: 13 of 14 files reviewed at latest revision, 11 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14966#-:-KiQdfm0RqNq7s6Sdq_-:b2vwz0c)*
<!-- Sent from Reviewable.io -->

petermattis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

@a-robinson In 1.1, I think we'll want to switch to storing the lease history persistently. 

---

Review status: 13 of 14 files reviewed at latest revision, 11 unresolved discussions, all commit checks successful.

---

*[pkg/storage/lease_history.go, line 25 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/14966#-KiS9jpz2JexkgNKUhcL:-KiS9jpz2JexkgNKUhcM:b-rksj4e) ([raw file](https://github.com/cockroachdb/cockroach/blob/0db95b2252be13535f7e4d0ca5f80691920e886c/pkg/storage/lease_history.go#L25)):*
> ```Go
> // leaseHistoryMaxEntries controls if replica lease histories are enabled and
> // how much memory they take up. Turning them on will encour some memory usage.
> var leaseHistoryMaxEntries = envutil.EnvOrDefaultInt("COCKROACH_LEASE_HISTORY", 0)
> ```

Disabled by default? Can we make this 5? I think we can spare 400 bytes per replica until we have a way to persist the lease history.

---

*[pkg/storage/lease_history.go, line 73 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/14966#-KiSACspRqCEJy1cbrF1:-KiSACspRqCEJy1cbrF2:b-9glkun) ([raw file](https://github.com/cockroachdb/cockroach/blob/0db95b2252be13535f7e4d0ca5f80691920e886c/pkg/storage/lease_history.go#L73)):*
> ```Go
> 	copy(firstResult, first)
> 	copy(secondResult, second)
> 	return append(firstResult, secondResult...)
> ```

You're duplicating `first` and `second`, then performing another allocation to append `second` to `first`. The above can be replaced with:

```
return append(append([]roachpb.Lease(nil), first...), second...)
```

Or if you want to be more explicit and controlled about the memory usage:

```
result := make([]roachpb.Lease, len(first)+len(second))
copy(result, first)
copy(result[len(first):], second)
return result
```

---

*[pkg/storage/replica.go, line 101 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/14966#-KiBErCFFniIoUja6HQx:-KiSAkKlIU15RsuJU4ac:by2xizj) ([raw file](https://github.com/cockroachdb/cockroach/blob/39ca3da1d98208f7e034fbc2226903d09ee25c57/pkg/storage/replica.go#L101)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

Sure, done.
</blockquote></details>

I probably missed some discussion: why default to 0? Having some small amount of history (I suggest 5 elsewhere) would be helpful and have little memory impact (4 MB for 10k replicas).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14966#-:-KiSHn78mMdTvOweF_rv:b-5xx3nn)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
If this is defaulted to off, then I'll be turning it on for all of our test clusters.  But I would rather is have a little bit of history, so I think the suggestion of 5 leases per replica is a good one.  

---

Review status: 11 of 14 files reviewed at latest revision, 11 unresolved discussions, all commit checks successful.

---

*[pkg/storage/lease_history.go, line 25 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/14966#-KiS9jpz2JexkgNKUhcL:-KiV_0As4IQGlpWo7bX2:byf9p2y) ([raw file](https://github.com/cockroachdb/cockroach/blob/0db95b2252be13535f7e4d0ca5f80691920e886c/pkg/storage/lease_history.go#L25)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Disabled by default? Can we make this 5? I think we can spare 400 bytes per replica until we have a way to persist the lease history.
</blockquote></details>

SGTM  Done.

---

*[pkg/storage/lease_history.go, line 73 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/14966#-KiSACspRqCEJy1cbrF1:-KiV_9ZJgkfZRtlzfb-k:b6x6xys) ([raw file](https://github.com/cockroachdb/cockroach/blob/0db95b2252be13535f7e4d0ca5f80691920e886c/pkg/storage/lease_history.go#L73)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

You're duplicating `first` and `second`, then performing another allocation to append `second` to `first`. The above can be replaced with:

```
return append(append([]roachpb.Lease(nil), first...), second...)
```

Or if you want to be more explicit and controlled about the memory usage:

```
result := make([]roachpb.Lease, len(first)+len(second))
copy(result, first)
copy(result[len(first):], second)
return result
```
</blockquote></details>

Done.  Went with the more explicit option.

---

*[pkg/storage/replica.go, line 101 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/14966#-KiBErCFFniIoUja6HQx:-KiVSd7hlCyy7yq3g-hM:b-5eutil) ([raw file](https://github.com/cockroachdb/cockroach/blob/39ca3da1d98208f7e034fbc2226903d09ee25c57/pkg/storage/replica.go#L101)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

I probably missed some discussion: why default to 0? Having some small amount of history (I suggest 5 elsewhere) would be helpful and have little memory impact (4 MB for 10k replicas).
</blockquote></details>

I'm all for turning this on by default.  But there was a good amount of pushback.  I've set the default to 5 now.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14966#-:-KiVZ_DSdNsGh89A48Fc:b-pkcvnn)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
LGTM

---

Reviewed 3 of 3 files at r9.
Review status: all files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14966#-:-KiVzsLUyWM2WNSmPnoO:bb6ywmx)*
<!-- Sent from Reviewable.io -->
